packages/flutter_tools/lib/src/web/compiler_config.dart

```diff
@@ -0,0 +1,137 @@
+          sourceMaps: true,
+        );
+
+  /// Creates a new [JsCompilerConfig] from build system environment values.
+  ///
+  /// Should correspond exactly with [toBuildSystemEnvironment].
+  factory JsCompilerConfig.fromBuildSystemEnvironment(
+          Map<String, String> defines) =>
+      JsCompilerConfig(
+        csp: defines[kCspMode] == 'true',
+        dumpInfo: defines[kDart2jsDumpInfo] == 'true',
+        nativeNullAssertions: defines[kNativeNullAssertions] == 'true',
+        optimizationLevel: defines[kDart2jsOptimization] ?? kDart2jsDefaultOptimizationLevel,
+        noFrequencyBasedMinification: defines[kDart2jsNoFrequencyBasedMinification] == 'true',
+        sourceMaps: defines[kSourceMapsEnabled] == 'true',
+      );
+
+  /// The default optimization level for dart2js.
+  ///
+  /// Maps to [kDart2jsOptimization].
+  static const String kDart2jsDefaultOptimizationLevel = 'O4';
+
+  /// Build environment flag for [optimizationLevel].
+  static const String kDart2jsOptimization = 'Dart2jsOptimization';
+
+  /// Build environment flag for [dumpInfo].
+  static const String kDart2jsDumpInfo = 'Dart2jsDumpInfo';
+
+  /// Build environment flag for [noFrequencyBasedMinification].
+  static const String kDart2jsNoFrequencyBasedMinification =
+      'Dart2jsNoFrequencyBasedMinification';
+
+  /// Build environment flag for [csp].
+  static const String kCspMode = 'cspMode';
+
+  /// Build environment flag for [sourceMaps].
+  static const String kSourceMapsEnabled = 'SourceMaps';
+
+  /// Build environment flag for [nativeNullAssertions].
+  static const String kNativeNullAssertions = 'NativeNullAssertions';
+
+  /// Whether to disable dynamic generation code to satisfy csp policies.
+  final bool csp;
+
+  /// If `--dump-info` should be passed to dart2js.
+  final bool dumpInfo;
+
+  /// Whether the dart2js native null assertions are enabled.
+  final bool nativeNullAssertions;
+
+  // If `--no-frequency-based-minification` should be passed to dart2js
+  // TODO(kevmoo): consider renaming this to be "positive". Double negatives are confusing.
+  final bool noFrequencyBasedMinification;
+
+  /// An override for the dart2js build mode.
+  ///
+  /// Valid values are O1 (lowest, profile default) to O4 (highest, release default).
+  // TODO(kevmoo): consider storing this as an [int] and validating it!
+  final String optimizationLevel;
+
+  /// Whether the dart2js build should output source maps.
+  final bool sourceMaps;
+
+  @override
+  bool get isWasm => false;
+
+  @override
+  Map<String, String> toBuildSystemEnvironment() => <String, String>{
+        kCspMode: csp.toString(),
+        kDart2jsDumpInfo: dumpInfo.toString(),
+        kNativeNullAssertions: nativeNullAssertions.toString(),
+        kDart2jsNoFrequencyBasedMinification: noFrequencyBasedMinification.toString(),
+        kDart2jsOptimization: optimizationLevel,
+        kSourceMapsEnabled: sourceMaps.toString(),
+      };
```

Comment:
```suggestion
  Map<String, String> toBuildSystemEnvironment() => <String, String>{
    kCspMode: csp.toString(),
    kDart2jsDumpInfo: dumpInfo.toString(),
    kNativeNullAssertions: nativeNullAssertions.toString(),
    kDart2jsNoFrequencyBasedMinification: noFrequencyBasedMinification.toString(),
    kDart2jsOptimization: optimizationLevel,
    kSourceMapsEnabled: sourceMaps.toString(),
  };
```

Reply:
Let's be thoughtful about where we fight dartfmt, right?